### PR TITLE
fix(get-function-name-with-kind): fix default exported function name

### PR DESCRIPTION
**Pull request by fisker**
Fixes https://github.com/mysticatea/eslint-utils/issues/24

### CODE REVIEW

**Comment by MichaelDeBoey** [comment]
@fisker Can you please split this PR into 2 separate commits?
One commit with failing CI tests & afterwards the fix?

This will make it easier to see how the problem was surfacing before the fix

**Comment by MichaelDeBoey** [comment]
Without the fix, added tests were failing like:

```sh
  1) The 'getFunctionNameWithKind' function
       should return "async generator function 'default'" for "export default async function* () {}".:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'async generator function'
- "async generator function 'default'"
      + expected - actual

      -async generator function
      +async generator function 'default'
      
      at Context.<anonymous> (test/get-function-name-with-kind.js:159:20)
      at processImmediate (node:internal/timers:464:21)

  2) The 'getFunctionNameWithKind' function
       should return "async generator function 'default'" for "export default async function* () {}" if sourceCode is present.:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'async generator function'
- "async generator function 'default'"
      + expected - actual

      -async generator function
      +async generator function 'default'
      
      at Context.<anonymous> (test/get-function-name-with-kind.js:189:20)
      at processImmediate (node:internal/timers:464:21)

  3) The 'getFunctionNameWithKind' function
       should return "async arrow function 'default'" for "export default async () => {}".:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'async arrow function'
- "async arrow function 'default'"
      + expected - actual

      -async arrow function
      +async arrow function 'default'
      
      at Context.<anonymous> (test/get-function-name-with-kind.js:159:20)
      at processImmediate (node:internal/timers:464:21)

  4) The 'getFunctionNameWithKind' function
       should return "async arrow function 'default'" for "export default async () => {}" if sourceCode is present.:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'async arrow function'
- "async arrow function 'default'"
      + expected - actual

      -async arrow function
      +async arrow function 'default'
```

File: test/get-function-name-with-kind.js [hunk 1]
```diff
@@ -64,6 +64,11 @@ describe("The 'getFunctionNameWithKind' function", () => {
         "class A { static async foo() {} }": "static async method 'foo'",
         "class A { static get foo() {} }": "static getter 'foo'",
         "class A { static set foo(a) {} }": "static setter 'foo'",
+        "export default async function* foo() {}":
+            "async generator function 'foo'",
+        "export default async function* () {}":
+            "async generator function 'default'",
+        "export default async () => {}": "async arrow function 'default'",
```
Comment (MichaelDeBoey) [inline test/get-function-name-with-kind.js:71]: It's normally the name of the function that we put in between quotes (`''`).
`default` isn't the name of the function, it's the type of export that you're doing.
The function is an anonymous function, so it has no name.

I think we should get it like 🤔

```suggestion
        "export default async function* () {}":
            "default async generator function",
        "export default async () => {}": "default async arrow function",
```

Comment (fisker) [reply test/get-function-name-with-kind.js:71]: `default` is the function name.

We do the same to 

```js
a.b = function() {} // -> `function 'b'`
```

Comment (fisker) [reply test/get-function-name-with-kind.js:71]:
```console
>node
Welcome to Node.js v18.12.1.
Type ".help" for more information.
> (await import("data:text/javascript,export default function(){}")).default.name
'default'
>
```

**Review by ota-meshi — APPROVED** [review]
Thank you!

**Pull request merged**